### PR TITLE
Fix Bulksettings Problem Settings page linking

### PR DESCRIPTION
**Pull request by albertliangcode**
On Bulksettings Problem Settings page, links redirect to these destinations:
- Section link > CourseOutline_Page/Section
- Subsection link > CourseOutline_Page/Section/Subsection
- Unit link > Unit_Page
- Problem link > Unit_Page

Clicking on a Section link sends the user to the Course Outline page, with the
page automatically scrolling down to the requested section.  Clicking on a
Subsection link sends the user to the Course Outline page, with the page
automatically scrolling down to the requested subsection.  Clicking on a Unit
link sends the user to the requested Unit page.  Clicking on a Problem link
sends the user to the Unit page where the requested problem is.

The get_settings_url_for_category method in utils.py no longer uses
component_handler for URL generation.  Instead, the method generates the URLs
itself using XBlock attributes and the CMS_BASE attribute defined in common.py.

TODO: Clicking on Problem link should send the user to the Unit page, with the
page automatically scrolling down to the requested problem.  The issue is
probably because the problems render after the HTML in JavaScript.  Using
JavaScript, there should be a way to make the page scroll down to the requested
problem. However, if the user has a slow internet connection, this
implementation may cause the page to jerk downwards long after the HTML has
rendered, possibly surprising the user.

### CODE REVIEW

**Comment by albertliangcode** [comment]
@stvstnfrd :  Whenever you're available, please review!  Have a great weekend!

**Comment by albertliangcode** [comment]
Fixing errors reported in Jenkins now.

**Comment by albertliangcode** [comment]
Both commits failed from the same lms error:

1) FAIL: test_unreleased_course (lms.djangoapps.mobile_api.course_info.tests.TestHandouts)

   Traceback (most recent call last):
    /var/lib/jenkins/workspace/Stanford_Online_Unit_Tests_auto_pr@2/venv/local/lib/python2.7/site-packages/mock.py line 1618 in _inner
      return f(*args, **kw)
    lms/djangoapps/mobile_api/testutils.py line 169 in test_unreleased_course
      self.verify_failure(response)
    lms/djangoapps/mobile_api/testutils.py line 142 in verify_failure
      self.assertEqual(response.status_code, 404)
   AssertionError: 200 != 404

Did I break a test for mobile?

File: cms/djangoapps/contentstore/utils.py [hunk 1]
```diff
@@ -396,16 +396,34 @@ def get_settings_url_for_category(cls, category, child, parent):
             - Problems: Unit url
         """
         handler = category
```
Comment (dcadams) [inline cms/djangoapps/contentstore/utils.py]: Maybe remove this line and use "category" throughout.

File: cms/djangoapps/contentstore/utils.py [hunk 2]
```diff
@@ -396,16 +396,34 @@ def get_settings_url_for_category(cls, category, child, parent):
...
-        if category == "chapter":
-            usage_key_string = unicode(parent.id)
-        elif category == "sequential" or category == "unit":
-            usage_key_string = unicode(child.location)
         else:
-            usage_key_string = unicode(parent.location)
-            handler = 'unit'
-
-        return reverse('component_handler', kwargs={'usage_key_string': usage_key_string, 'handler': handler})
+            assertEqual(handler, 'problem')
```
Comment (dcadams) [inline cms/djangoapps/contentstore/utils.py]: What's the purpose of this line?

**Comment by dcadams** [comment]
If there wasn't any existing unit tests for this then you should craft some. Take a look at code coverage to determine.

File: cms/djangoapps/contentstore/utils.py (hunk 2 above)
Comment (albertliangcode) [inline cms/djangoapps/contentstore/utils.py]: I thought it would add a little extra security and clarity - the original code set any 'category' or 'handler' that wasn't a 'chapter', 'sequential', or 'unit' to be a 'unit'.  Unfortunately, this meant that anything else, including 'problem' objects, would be handled in the else statement, even though they were returned as 'unit' objects.  I thought it was confusing, and tried to change things in the second commit of this pull request.  I hoped that this code would make sure that any object that failed the if and elif statements was a 'problem' and not something weird.

If this is unnecessary, I can remove this second commit right away!

Comment (albertliangcode) [inline cms/djangoapps/contentstore/utils.py]: Okay!  Will remove assertEqual and 'handler' variable, raise an exception to handle edge case, and refactor code so the 'url+=' is not as repetitive.  Will ask later about designing good unit tests.

**Comment by albertliangcode** [comment]
test this please

**Comment by albertliangcode** [comment]
Discovered that previous lms test error was the result of a flaky, outdated test.  The test has been skipped upstream - we can ignore the test for now.

**Comment by albertliangcode** [comment]
Well, thank you for your effort, @stanford-online-robot .  You responded very quickly to test this please.

**Comment by albertliangcode** [comment]
@stvstnfrd : @caesar2164 helped me find an XBlock method that takes care of everything in a much cleaner way.  I've refactored the code, though it might mean I'd have to ask you to review my code all over again.  After I finish the tests, is it okay for me to push the changes to this PR?

**Comment by albertliangcode** [comment]
@stanford-online-robot !

**Comment by albertliangcode** [comment]
Robot, I didn't want you to test this please.

**Comment by albertliangcode** [comment]
@stvstnfrd : Everything looks like it checks out.  Paver returns no errors on the refactored code.  I've also rebased the branch to include Giulio's cherry-pick.  I can push the code anytime!

**Comment by albertliangcode** [comment]
@stvstnfrd :  I just wanted to be cautious in case you had started writing comments - I didn't want to bulldoze your feedback into outdated diffs.  If you're not on GitHub right now, I'll go ahead and push the changes.

**Comment by stvstnfrd** [comment]
@albertliangcode 

> Robot, I didn't want you to test this please.

Any time you push changes to a branch w/ an open PR, it will rerun tests. So once a PR is open, only push changes once you're ready for a test build to be run.

**Comment by albertliangcode** [comment]
Sorry, I just wanted to see if the robot would respond to the testing string if it was buried in a sentence.  Thank you for letting me know, though!

**Comment by stvstnfrd** [comment]
Thanks for taking the time to walk through git-surgery with me yesterday; this is much easier to review now!

I have completed my first pass. I have a few suggestions, as well as a couple of clarifying questions.

**Comment by albertliangcode** [comment]
Thank you for walking me through git-surgery, and for reviewing my code @stvstnfrd !  I made the changes - hopefully everything is still easy to read.  Please review this again!

File: cms/djangoapps/contentstore/views/utilities/bulksettings.py [hunk 3]
```diff
@@ -9,14 +9,113 @@
...
-from contentstore.utils import BulkSettingsUtil
-
+from contentstore.views.helpers import xblock_studio_url
 from opaque_keys.edx.keys import CourseKey
 from xmodule.modulestore.django import modulestore
 
 from student.auth import has_course_author_access
 from ..course import get_course_and_check_access
 
+from datetime import datetime
```
Comment (stvstnfrd) [inline cms/djangoapps/contentstore/views/utilities/bulksettings.py]: Sorry for the confusion; I must have gotten mixed up in the commits. This import _should_ appear in the top stanza (alongside `logging` since they're both part of the standard library, but alphabetized).

This change should be part of the initial commit where we move the class definition, so the easiest way to address that is to:
- add a new commit that puts in in the right place
- rebase to reorder and squash the new commit into the first
- resolve any conflicts by keeping the `import` in the proper location.

**Comment by stvstnfrd** [comment]
Hey @albertliangcode I just had one other quick comment. After that, can you give me a quick walkthrough?

**Comment by albertliangcode** [comment]
Ready when you are, @stvstnfrd !

**Comment by albertliangcode** [comment]
@stvstnfrd : Finished packaging PR!  Please let me know if it's okay to merge!

**Comment by stvstnfrd** [comment]
Very clear commit messages!

+1 once tests all pass

**Comment by albertliangcode** [comment]
Merged, and remote branches cleaned!

**Comment by stvstnfrd** [comment]
w00t

Nice work!
